examples_tests/13.BRDF_Explorer/BRDFExplorerApp.cpp

```diff
@@ -340,54 +340,69 @@ float applyChainRule(in vec2 h_gradient, in vec2 dUVd_)
 }
 
 // Calculate the surface normal using the uv-space gradient (dhdu, dhdv)
-vec3 calculateSurfaceNormal(in vec3 position, in vec2 uv, in vec3 normal, in vec2 h_gradient)
+vec3 calculateSurfaceNormal(in vec3 normal, in vec2 h_gradient, in vec3 dpdx, in vec3 dpdy, in vec4 dUVd_)
 {
-    vec3 dpdx = dFdx(position);
-    vec3 dpdy = dFdy(position);
- 
-	vec2 dUVdx = dFdx(uv);
-	vec2 dUVdy = dFdy(uv);
+	vec2 dUVdx = dUVd_.xy;
+	vec2 dUVdy = dUVd_.zw;
  
     float dhdx = applyChainRule(h_gradient, dUVdx);
     float dhdy = applyChainRule(h_gradient, dUVdy);
  
     return perturbNormal(normal, dpdx, dpdy, dhdx, dhdy);
 }
 
+float calcDiffuseCorrectionFactor_(float n)
```

Comment:
you know you could just do this whole function in `vec3` then if someone wants the `float` version then they just put `.x` at the end ;)

also call it `calcDiffuseFresnelCorrectionFactor`

Reply:
i cant because theres branch and i have no guarantee all of them will be on the same side of 1

Reply:
for such short functions you should aim for branchless (every branch generates at least 3 instructions + each branch could be executed, a mask generates 1), also you have eta2 laying around (by necessity you alwasy have it, so use it).

Yes DIV is hundreds of cycles, but I'm hoping for a reciprocal instruction to be generated.

Here's an implementation (check it)
```cpp
float calcDiffuseCorrectionFactor(in float n, in float n2)
{
   //assert(n*n==n2);
   bvec3 TIR = n<1.0;
   float invdenum = mix(1.0,1.0/(n2*n2*(554.33 - 380.7*n)),TIR);
   float num = n*mix(0.1921156102251088,n*298.25 - 261.38*n2 + 138.43,TIR);
   num += mix(0.8078843897748912,-1.67,TIR);
   return
}
```